sweyer_plugin/android/src/main/kotlin/com/nt4f04und/sweyer/sweyer_plugin/handlers/FetchHandler.kt

```diff
@@ -0,0 +1,233 @@
+            MediaStore.Audio.Artists.EXTERNAL_CONTENT_URI,
+            MediaStore.Audio.Artists.ARTIST + " IS NOT NULL",
+            mapOf(
+                MediaStore.Audio.Artists._ID to Pair("id") { cursor, index -> cursor.getInt(index) },
+                MediaStore.Audio.Artists.ARTIST to Pair("artist") { cursor, index -> cursor.getString(index) },
+                MediaStore.Audio.Artists.NUMBER_OF_ALBUMS to Pair("numberOfAlbums") { cursor, index ->
+                    cursor.getInt(index)
+                },
+                MediaStore.Audio.Artists.NUMBER_OF_TRACKS to Pair("numberOfTracks") { cursor, index ->
+                    cursor.getInt(index)
+                },
+            )
+        )
+    }
+
+    fun retrieveGenres(resolver: ContentResolver): ArrayList<MutableMap<String, Any?>> {
+        val memberProjection = arrayOf(MediaStore.Audio.Genres.Members._ID)
+        return executeQuery(
+            resolver,
+            MediaStore.Audio.Genres.EXTERNAL_CONTENT_URI,
+            MediaStore.Audio.Genres.NAME + " IS NOT NULL",
+            mapOf(
+                MediaStore.Audio.Genres._ID to Pair("id") { cursor, index -> cursor.getInt(index) },
+                MediaStore.Audio.Genres.NAME to Pair("name") { cursor, index -> cursor.getString(index) },
+            )
+        ).onEach { genreInfo ->
+            val id = genreInfo["id"] as Int
+            resolver.query(
+                MediaStore.Audio.Genres.Members.getContentUri("external", id.toLong()),
+                memberProjection,
+                null,
+                null,
+                null
+            )?.use { membersCursor ->
+                val songIds = ArrayList<Int>()
+                while (membersCursor.moveToNext()) {
+                    songIds.add(membersCursor.getInt(0))
+                }
+                genreInfo["songIds"] = songIds
+            }
+        }
+    }
+
+    /**
+     * Execute the [query] against the [uri] in [resolver].
+     * The [fields] map describes each field in the resulting map:
+     * The key is the name of the field in the queried database,
+     * the value is the key in the new map and a function to extract the value from the cursor by its index.
+     */
+    private fun executeQuery(
+        resolver: ContentResolver, uri: Uri, query: String, fields: Map<String, Pair<String, (Cursor, Int) -> Any?>>
+    ): ArrayList<MutableMap<String, Any?>> {
+        val maps = ArrayList<MutableMap<String, Any?>>()
+        resolver.query(uri, fields.keys.toTypedArray(), query, null, null)?.use { cursor ->
+            val columns = fields.keys.map { cursor.getColumnIndexOrThrow(it) }
+            while (cursor.moveToNext()) {
+                val itemInfo = mutableMapOf<String, Any?>()
+                fields.values.zip(columns).associateTo(itemInfo) { (fieldInfo, index) ->
+                    val (name, accessor) = fieldInfo
+                    name to if (cursor.isNull(index)) null else accessor(cursor, index)
```

Comment:
The `cursor.isNull` check here is also a major difference. According to the [`getString` doccumentation](https://developer.android.com/reference/android/database/Cursor#getString(int)):

> The result and whether this method throws an exception when the column value is null or the column type is not a string type is implementation-defined.

Same goes for `getInt`. It's better to check for null on every field in order to avoid throwing an exception.
I noticed this because the return value of `getString` is not marked as nullable, which causes Kotlin to generate an implicit null check and throw a `NullPointerException`.